File: crates/klyron_runtime/bindings/cpp/builder.cpp
#include "builder.hpp"

namespace klyron {

class Klyron::RuntimeBuilder::Impl {
public:
  Klyron::RuntimeConfig config;
  bool verbose = false;
};

Klyron::RuntimeBuilder::Klyron::RuntimeBuilder()
  : impl_(std::make_unique<Impl>()) {
}

Klyron::RuntimeBuilder& Klyron::RuntimeBuilder::withConfig(const Klyron::RuntimeConfig& config) {
  impl_->config = config;
  return *this;
}

Klyron::RuntimeBuilder& Klyron::RuntimeBuilder::verbose(bool v) {
  impl_->verbose = v;
  return *this;
}

Klyron::RuntimeBuilder::Klyron::RuntimeInstance Klyron::RuntimeBuilder::build() {
  Klyron::RuntimeInstance inst;
  inst.config = impl_->config;
  inst.verbose = impl_->verbose;
  return inst;
}

} // namespace klyron

File: crates/klyron_runtime/bindings/rust/builder.rs
//! Builder pattern for klyron_runtime

use crate::types::Klyron::RuntimeConfig;

#[derive(Debug, Default)]
pub struct Klyron::RuntimeBuilder {
    config: Option<Klyron::RuntimeConfig>,
    verbose: bool,
}

impl Klyron::RuntimeBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_config(mut self, config: Klyron::RuntimeConfig) -> Self {
        self.config = Some(config);
        self
    }

    pub fn verbose(mut self, verbose: bool) -> Self {
        self.verbose = verbose;
        self
    }

    pub fn build(self) -> anyhow::Result<Klyron::RuntimeInstance> {
        let config = self.config.unwrap_or_default();
        Ok(Klyron::RuntimeInstance { config, verbose: self.verbose })
    }
}

#[derive(Debug)]
pub struct Klyron::RuntimeInstance {
    pub config: Klyron::RuntimeConfig,
    pub verbose: bool,
}
